File: app/src/main/java/ru/laushkina/rates/ui/RatesActivity.java
package ru.laushkina.rates.ui;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.util.Pair;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.work.WorkManager;

import java.util.List;

import io.reactivex.disposables.Disposable;
import ru.laushkina.rates.R;
import ru.laushkina.rates.model.RatesService;
import ru.laushkina.rates.model.RateShortName;
import ru.laushkina.rates.repository.RatesDbRepository;
import ru.laushkina.rates.util.RatesLog;

public class RatesActivity extends Activity implements RatesView, RateAdapter.ValueChangeListener {
    private static final String TAG = "RatesActivity";
    private RecyclerView ratesRecycler;
    private RatesPresenter ratesPresenter;
    private Disposable itemClickDisposable;
    private RateAdapter rateAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_rates);

        ratesRecycler = findViewById(R.id.rates);

        RatesService service = new RatesService(new RatesDbRepository(getAppContext()));
        ratesPresenter = new RatesPresenter(service, WorkManager.getInstance(getAppContext()), this);
        ratesPresenter.onCreate();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        ratesPresenter.onDestroy();
        if (itemClickDisposable != null) {
            itemClickDisposable.dispose();
        }
    }

    @Override
    public void showRates(@NonNull List<RateViewModel> rates) {
        if (rateAdapter == null) {
            rateAdapter = new RateAdapter(rates, this);
            itemClickDisposable = rateAdapter.getPositionClicks().subscribe(viewModel -> {
                ratesPresenter.onRateSelected(viewModel);
                ratesRecycler.scrollToPosition(0);
            });

            ratesRecycler.setAdapter(rateAdapter);
            ratesRecycler.setLayoutManager(new GridLayoutManager(this, 1));
            ratesRecycler.setHasFixedSize(true);
        } else {
            ratesRecycler.post(() -> {
                rateAdapter.updateRates(rates);
            });
        }
    }

    @Override
    public void showError(@NonNull Throwable throwable) {
        RatesLog.e(TAG, "", throwable);
        Toast.makeText(this, "Cannot update rates: " + throwable.getMessage(), Toast.LENGTH_LONG).show();
    }

    @NonNull
    @Override
    public Context getAppContext() {
        return getApplicationContext();
    }

    @Override
    public void beforeValueChange(@NonNull RateViewModel rate) {
        ratesPresenter.beforeRateValueChange();
    }

    @Override
    public void afterValueChange(@NonNull RateViewModel rate) {
        ratesPresenter.afterRateValueChange();
    }

    @Override
    public void onValueChange(@NonNull RateViewModel rate, @NonNull String value) {
        ratesPresenter.onRateValueChange(rate, value);
    }
}

File: app/src/main/java/ru/laushkina/rates/ui/RatesActivity.kt
package ru.laushkina.rates.ui

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.widget.Toast
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import ru.laushkina.rates.ContextModule
import ru.laushkina.rates.DaggerApplicationComponent
import ru.laushkina.rates.R
import ru.laushkina.rates.RatesViewModule
import ru.laushkina.rates.util.RatesLog
import javax.inject.Inject

class RatesActivity : Activity(), RatesView, RateAdapter.ValueChangeListener, RateAdapter.RateClickListener {
    companion object {
        private const val TAG = "RatesActivity"
    }

    private lateinit var ratesRecycler: RecyclerView
    @Inject lateinit var ratesPresenter: RatesPresenter
    private var rateAdapter: RateAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_rates)
        ratesRecycler = findViewById(R.id.rates)

        ratesPresenter = DaggerApplicationComponent.builder()
                .contextModule(ContextModule(this))
                .ratesViewModule(RatesViewModule(this))
                .build().getRatesPresenter()

        ratesPresenter.onCreate()

        val updateFab: FloatingActionButton = findViewById(R.id.update_fab)
        updateFab.setOnClickListener { ratesPresenter.onUpdateRequested() }
    }

    override fun onDestroy() {
        super.onDestroy()
        ratesPresenter.onDestroy()
    }

    override fun getAppContext(): Context {
        return applicationContext
    }

    override fun showRates(rates: MutableList<RateViewModel>) {
        if (rateAdapter == null) {
            rateAdapter = RateAdapter(rates, this, this)
            ratesRecycler.adapter = rateAdapter
            ratesRecycler.layoutManager = GridLayoutManager(this, 1)
            ratesRecycler.setHasFixedSize(true)
        } else {
            ratesRecycler.post { rateAdapter!!.updateRates(rates) }
        }
    }

    override fun showError(throwable: Throwable) {
        RatesLog.e(TAG, "Common error", throwable)
        Toast.makeText(this, "Cannot update rates: " + throwable.message, Toast.LENGTH_LONG).show()
    }

    override fun showTooEarlyForNewUpdate() {
        Toast.makeText(this, "Too early for new update", Toast.LENGTH_LONG).show()
    }

    override fun onValueChange(rate: RateViewModel, value: String) {
        ratesPresenter.onRateValueChange(rate, value)
    }

    override fun onClicked(position: Int, rate: RateViewModel) {
        ratesRecycler.scrollToPosition(0)
        ratesPresenter.onRateSelected(rate)
    }
}